### ISSUE TITLE
Complex types support

### ISSUE THREAD

**Issue by 4nonym0us** [gtaylor44/SqlBulkTools#33]
Hello, I've spend a couple of days exploring SqlBulkTools and I really enjoy it! Thanks a lot for supporting the project.
As far as I can see ComplexTypes aren't being mapped automatically (via `AddAllColumns`) and they cannot be added manually (via `AddColumn`) because SqlBulkSetup uses CompleType's property name as Id (therefore, only 1 property of complex type can be mapped . Are there any plans of getting ComplexTypes working with SqlBulkTools or am I doing anything wrong?

**Comment by gtaylor44** [comment]
Thanks for your kind feedback. ComplexTypes are not currently supported so rest assured you're not doing anything wrong. Currently, types that are not value or string type (and a few other types) are being ignored from AddAllColumns and throw an exception with AddColumn because they're not compatible with SQL Server types. 

Could you show me an example of your model that you're trying to work with?

**Comment by 4nonym0us** [comment]
The list of my entity properties is below, the rest code (class ctors, methods, etc.) is omitted for brevity:
```csharp

        public long KeywordId { get; set; }

        public EstimatedStats MinEstimate { get; set; }

        public EstimatedStats AverageEstimate { get; set; }

        public EstimatedStats MaxEstimate { get; set; }

        public double SearchVolume { get; set; }

        public double Competition { get; set; }

        public double AdjustedSearchVolume => SearchVolume*(1 - Competition);

        public DateTime CreationTime { get; set; }
```
Where `EstimatedStats` is a complex type:
```csharp

    [ComplexType]
    public class EstimatedStats
    {
        public EstimatedStats()
        {
            CreationTime = Clock.Now;
        }

        public double? AverageCpc { get; set; }

        public double? AveragePosition { get; set; }

        public double? ClickThroughRate { get; set; }

        public float? ClicksPerDay { get; set; }

        public float? ImpressionsPerDay { get; set; }

        public double? TotalCost { get; set; }

        public DateTime CreationTime { get; set; }
    }
```
By the way, take a loot at property `AdjustedSearchVolume`. It's computed (computed on run time, since there it is no [Compured] flag) property and it shouldn't normally included during AddAllColumns. I was getting an mapping error until I removed it manually from BulkSetup. I don't know if this behavior is intended, but It looks like a bug.

**Comment by gtaylor44** [comment]
Thanks for that.

By the way.. If using AddAllColumns you can use RemoveColumn(x => x.AdjustedSearchVolume) afterwards to manually exclude it.

**Comment by gtaylor44** [comment]
Is it possible to also see an example schema of your target table(s)? How is EstimatedStats structured?

**Comment by 4nonym0us** [comment]
Yeah, I've read the docs and used `RemoveColumn` to remove the column and fix the issue.
Fluent api BulkSetup creation works extremely well and it's easy-to-use.
Do you need SQL-dump or a snapshot with columns/DataTypes-DbTypes would be enough?

**Comment by gtaylor44** [comment]
Just a snapshot of columns w/ types should be enough.

**Comment by 4nonym0us** [comment]
Snapshot:
![image](https://cloud.githubusercontent.com/assets/2701409/24137270/b914557a-0e1b-11e7-88a7-64df87a5c0a4.png)
Already had enough time to run the SSMS so I can share the table creation snippet as well:
```sql

CREATE TABLE [dbo].[AppKeywordEstimates](
	[Id] [bigint] IDENTITY(1,1) NOT NULL,
	[KeywordId] [bigint] NOT NULL,
	[MinEstimate_AverageCpc] [float] NULL,
	[MinEstimate_AveragePosition] [float] NULL,
	[MinEstimate_ClickThroughRate] [float] NULL,
	[MinEstimate_ClicksPerDay] [real] NULL,
	[MinEstimate_ImpressionsPerDay] [real] NULL,
	[MinEstimate_TotalCost] [float] NULL,
	[MinEstimate_CreationTime] [datetime] NOT NULL,
	[AverageEstimate_AverageCpc] [float] NULL,
	[AverageEstimate_AveragePosition] [float] NULL,
	[AverageEstimate_ClickThroughRate] [float] NULL,
	[AverageEstimate_ClicksPerDay] [real] NULL,
	[AverageEstimate_ImpressionsPerDay] [real] NULL,
	[AverageEstimate_TotalCost] [float] NULL,
	[AverageEstimate_CreationTime] [datetime] NOT NULL,
	[MaxEstimate_AverageCpc] [float] NULL,
	[MaxEstimate_AveragePosition] [float] NULL,
	[MaxEstimate_ClickThroughRate] [float] NULL,
	[MaxEstimate_ClicksPerDay] [real] NULL,
	[MaxEstimate_ImpressionsPerDay] [real] NULL,
	[MaxEstimate_TotalCost] [float] NULL,
	[MaxEstimate_CreationTime] [datetime] NOT NULL,
	[SearchVolume] [float] NOT NULL,
	[Competition] [float] NOT NULL,
	[CreationTime] [datetime] NOT NULL,
 CONSTRAINT [PK_dbo.AppKeywordEstimates] PRIMARY KEY CLUSTERED 
(
	[Id] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON)
)
```

**Comment by gtaylor44** [comment]
Ahh I see.. Prefixed with the property name. Is this how EntityFramework always deals with ComplexType?

**Comment by 4nonym0us** [comment]
Yeah, that's how EF stores every ComplexProperty.

**Comment by gtaylor44** [comment]
Ok sounds good, I will try to get this in as soon as I can.

**Comment by 4nonym0us** [comment]
Great, thanks for a fast feedback!

**Comment by gtaylor44** [comment]
I'm working on something similar and have some questions (unrelated to this issue) to ask. Are you free to chat some time? My skype is greg.taylor4

**Comment by gtaylor44** [comment]
If you need this feature urgently you can checkout the ComplexTypeFeature branch. Im doing final tests at the moment but everything is looking good. Im overseas until Monday with no access to the project. Otherwise I will merge into master and release when I get back. EstimatedStats (or other ComplexType) must always be decorated with ComplexType attribute to be compatible.

**Comment by 4nonym0us** [comment]
Yeah, I saw that you has been developing complex types support on alternate branch, thanks a lot for spending your time on this issue. It's not extremely urgent as I've managed to create an temporary bug fix to support complex types (it looks ugly but it works if you create  expression-bodied properties to proxy the complex types' properties).
So I'm looking forward to see the complex types but there is no need to hurry: take your time and have a good time overseas.
